risc0/core/src/field/baby_bear.rs

```diff
@@ -681,14 +695,11 @@ impl ops::MulAssign for ExtElem {
         //             = c_i * sum_j=0^(EXT_SIZE - 1) -IRREDUCIBLE_WITHOUT_LEADING_TERM * x^(i - EXT_SIZE + j)
         // So, we can zero out the x^i term by adding c[i] * -IRREDUCIBLE[j] to c[i - EXT_SIZE + j] for j < EXT_SIZE
         assert_eq!(IRREDUCIBLE.len(), EXT_SIZE + 1);
-        assert_eq!(IRREDUCIBLE[EXT_SIZE], Elem::new(1));
-        let upper = 2 * EXT_SIZE - 2;
-        for i_rev in (0..EXT_SIZE - 1).rev() {
-            // We need to iterate from high degree to low, otherwise we might add to coefficients we've already zeroed out
-            let i = upper - i_rev;
-            for j in 0..(IRREDUCIBLE.len() - 1) {
-                // Perhaps there's a minor perf gain to be had by storing NEG_IRREDUCIBLE instead of IRREDUCIBLE itself
-                c[i - EXT_SIZE + j] += c[i] * -IRREDUCIBLE[j];
+        assert_eq!(IRREDUCIBLE[EXT_SIZE], Elem::ONE);
+        // We need to iterate from high degree to low, otherwise we might add to coefficients we've already zeroed out.
+        for i in EXT_SIZE..2 * EXT_SIZE - 1 {
+            for j in 0..EXT_SIZE {
+                c[i - EXT_SIZE + j] += c[i] * NEG_IRREDUCIBLE[j];
             }
```

Comment:
Thinking about my previous comment, iterating from low degree to high is legal in our case because in the inner `j` loop, `NEG_IRREDUCIBLE[j] = 0` for `j != 0`; thus,  because of the size of the `i` loop, the `c[i] = Elem::ZERO` of the outer loop never gets reset by the inner loop.

This _also_ means that under the same assumptions we're already making, we don't need the inner loop at all; I propose we either remove it, or revert the order of the `i` loop.
```suggestion
            // No loop is needed over the irred. poly. since it only has leading and constant terms
            c[i - EXT_SIZE] += c[i] * NEG_IRREDUCIBLE[0];
```

Reply:
But we are writing to the output from low to high, which is why I still consider it to be in reverse order. It's just that the index variable is going up.

Reply:
What happens if we select a different irreducible polynomial in the future? I'd like to keep the code generic so we don't have to change two things.

Reply:
> But we are writing to the output from high to low

One of us is missing something here, because I don't see this code writing the output from high to low: Noting that `j = 0` is the only iteration of the inner loop that writes anything (for our current irreducible polynomial), we write first to `c[0]`, then to `c[1]`, etc, and finally write to `c[EXT_SIZE - 1]` when `i == 2 * EXT_SIZE - 1`. Even if we were to use a more complicated polynomial, we would still be writing to increasing indices of `c` as `i` increased.

For keeping the code generic, I have two thoughts:
* It's a pretty simple change to keep it generic, especially with the inner loop already written.
* It's a kind of generic we're unlikely to ever use -- I'm pretty sure you can always find an irreducible polynomial of the form `x^EXT_SIZE + c`, so there's no need to use more complicated polynomials; and more complicated irreducible polynomials have minor performance disadvantages (right here, in fact: they require this inner `j` loop and `x^EXT_SIZE + c` polynomials don't).

Between those two considerations I'm ambivalent about whether to code for generic irreducible polynomials here. I do think we should design the full algorithm to be generic _or_ take advantage of every simplification available if we decide to not write code that's compatible with generic irreducible polynomials.

Reply:
Is there a good resource you can recommend on where to learn about polynomial reductions?

Reply:
Thanks, I'm going to copy paste this and remove from here because it doesn't seem like a good long term place for it and will hurt our ability to review the changes in the future. Generally we should treat PRs as a historical record for future developers.

Reply:
Good point, it definitely makes more sense somewhere long-term. Can you link it here for reference? (Or if you haven't made the Notion page yet I can do the porting using Git's history feature.)

Reply:
Hmm... mb I can do a study club on this; I can't think of any references on this topic that would be particularly helpful. 

For developing intuition on how familiar arithmetic algorithms translate into polynomial algorithms, I think James Tanton's course on exploding dots is phenomenal, but it's very non-traditional: https://gdaymath.com/lessons/explodingdots/6-1-welcome/

Reply:
> What happens if we select a different irreducible polynomial in the future? I'd like to keep the code generic so we don't have to change two things.

There are lots of algorithmic tricks that are specific to different choices of extension size and minimal polynomial. For example, using a minimal polynomial such as `x^2+x+1` or `x^6+x^3+1` will yield field extensions that have a "golden-ratio structure" that may offer a performance advantage over something like `x^5+B`. That said, I think for the foreseeable future we'll probably only be using extensions of the form `x^n+B`.

Reply:
> Can you link it here for reference?

Ya I'd love to take a look at this!

Reply:
https://www.notion.so/risczero/Extension-Fields-5e42931293894944ad164b3dc738e394?pvs=4